Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -2253,55 +2191,77 @@ void CGContextClearRect(CGContextRef context, CGRect rect) {
                                  nullptr);
     }
 
-    ComPtr<ID2D1Image> currentImage{ image };
+    auto cleanup = wil::ScopeExit([this, layer]() {
+        if (layer) {
+            this->deviceContext->PopLayer();
+        }
 
-    ComPtr<ID2D1Effect> shadowEffect;
-    RETURN_IF_FAILED(_CreateShadowEffect(currentImage.Get(), &shadowEffect));
-    if (shadowEffect) {
-        RETURN_IF_FAILED(shadowEffect.As(&currentImage));
-    }
+        // TODO GH#1194: We will need to re-evaluate Direct2D's D2DERR_RECREATE when we move to HW acceleration.
+        this->PopEndDraw();
+    });
+
+    if (state.HasShadow()) {
```

Comment:
I know I deferred you previously, but I still think this is out of scope? Currently, CGBitmapContexts that do not opt into PushBeginDraw()/PopEndDraw() get immediate mode rendering, as per reference platform. What you're asking me to do is to enable buffering for a specific subset of CGBitmapContexts, with caveats to make it conform to the expected behavior. However, I don't believe we've seen evidence that this subset is a significant perf concern, nor that your suggested optimization, with its caveats, would significantly impact such a bottleneck. As such, I still suggest we defer this recommendation until such a bottleneck actually arises.